### PR TITLE
[None][feat] Fully non-blocking pipeline parallelism executor loop.

### PR DESCRIPTION
**Pull request by yuxianq**
## Description

In pp executor loop, there are two kinds of data flows between ranks:

1. request flows: send via rank 0 -> rank 1 -> ... -> rank n-1, e.g., new request, scheduled batch.
2. response flows: send via rank n-1 -> rank 0 -> ... -> rank n-2, e.g., sample state.

We interlace request/response flows in the same executor loop.
Since we always wait sample state after n iteration since we submit its forward step, the last rank will block rank 0's executor loop if the forward steps of different ranks are uneven.

By executing response flows in a background thread, this PR enables a fully non-blocking pp executor loop.
In pp executor loop, we only handle these responses when available in `_handle_executed_batch`.
To keep the execution state of different ranks in sync, we count the available responses on rank 0 and send the response number to other ranks, so that all ranks handle the same set of responses in the same iteration.

There are two response flows in the previous implementation:

1. send host tensors of sample state
2. send logits if needed

In this PR, we merge these two response flows into a single flow, and send the diff of logits instead to reduce communication overhead.

This PR also contains the following improvements:

1. Add `TLLM_PP_ASYNC_BROADCAST_SAMPLE_STATE` to disable the non-blocking sample state broadcast, which is useful to get a deterministic result in unit tests or as fallback.
2. Allocate at least 1024 slots for inflight micro batches, which allows more uneven response intervals. It can be controlled by `MIN_ASYNC_MICRO_BATCH_NUM`.
3. Send all sample states via the same tag `PP_COMM_TAG_SAMPLE_STATE` to avoid tag conflict. So as `TERMINATION_COMM_TAG`.
4. TP broadcast scheduled batch among the first pp ranks to keep them in sync.
5. Add `tensor_to_str` for debugging.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 1]
```diff
@@ -115,6 +116,7 @@ class BatchStatePP(BatchState):
 
 
 class PyExecutor:
+    MIN_ASYNC_MICRO_BATCH_NUM = 1024
```
Comment (reasonsolo) [inline tensorrt_llm/_torch/pyexecutor/py_executor.py]: Will the queue exceed this max size causing some hang?

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 2]
```diff
@@ -1138,115 +1186,218 @@ def _executor_loop_pp(self):
...
-                            sample_state.host = self.dist.recv_object(
-                                src=self.dist.prev_pp_rank,
-                                tag=tag,
+                    self.executed_batch_queue.put(previous_batch)
+                    self.unhandled_batch_counter += 1
+
+                dequeue_counter = 0
+                executed_batch_num = 0
+
+                # The first rank determines the number of executed batches.
```
[the remark this answers is not in the document]
Comment (yuxianq) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py]: Good idea, I will create a follow-up PR for it. This PR is complicated enough.

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 3]
```diff
@@ -67,9 +67,10 @@
 PROFILE_TRACE_ENV_VAR_NAME = "TLLM_TORCH_PROFILE_TRACE"
 
 # Unique tag base to avoid collisions with token/logits comms
-TERMINATION_COMM_TAG_BASE = 20000
-PP_COMM_TAG_SCHEDULE_RESULT = 21000
-PP_COMM_TAG_SAMPLE_STATE_BASE = 21001
+TERMINATION_COMM_TAG = 20000
+PP_COMM_TAG_SCHEDULE_RESULT = 20001
+PP_COMM_TAG_EXECUTED_BATCH_NUM = 20002
+PP_COMM_TAG_SAMPLE_STATE = 20003
```
[the remark this answers is not in the document]
Comment (yuxianq) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py]: Sure, updated.

File: tensorrt_llm/_torch/pyexecutor/llm_request.py [hunk 4]
```diff
@@ -228,6 +228,25 @@ def set_log_probs(self, log_probs: list[TokenLogprobs],
 class PyResult:
     """PyResult reimplements some features of `bindings.executor.Result` in Python"""
 
+    @dataclass
+    class Diff:
```
[the remark this answers is not in the document]
Comment (yuxianq) [reply tensorrt_llm/_torch/pyexecutor/llm_request.py]: Added docstring for `Diff`.

**Review by reasonsolo — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 5]
```diff
@@ -1138,115 +1191,218 @@ def _executor_loop_pp(self):
...
+        while True:
+            previous_batch = self.executed_batch_queue.get()
+            if previous_batch is None:
+                break
+            self._broadcast_sample_state(previous_batch)
+        set_thread_local_mpi_comm(None)
+        new_mpi_comm.Free()
 
-                        self._remove_inflight_ids(previous_batch)
+    def _broadcast_sample_state(self, previous_batch: Optional[BatchStatePP]):
```
Comment (Funatiq) [inline tensorrt_llm/_torch/pyexecutor/py_executor.py]: I understand that it is logically a broadcast, but I'm wondering if we should make it more clear that we send along the pipeline ranks instead of 1-to-all communication. Should we rename it to something like `_communicate_sample_state_pp`?

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 6]
```diff
@@ -994,6 +1040,13 @@ def _executor_loop_pp(self):
                 # Fetch new requests from request queue
                 new_requests = self._fetch_and_activate_new_requests()
                 if self.should_stop_processing:
+                    while self.unhandled_batch_counter > 0:
+                        with nvtx_range("get_executed_batch"):
+                            executed_batch = self.executed_batch_response_queue.get(
+                            )
+                        self._handle_executed_batch(executed_batch)
+                        self.unhandled_batch_counter -= 1
```
Comment (Funatiq) [inline tensorrt_llm/_torch/pyexecutor/py_executor.py]: Should we move this outside of the `while True` loop? Not sure if we need to consider other `break` cases in the future. It would also be closer to the other `_handle_executed_batch` calls and might be easier to read.

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 7]
```diff
@@ -115,6 +121,7 @@ class BatchStatePP(BatchState):
 
 
 class PyExecutor:
+    MIN_ASYNC_MICRO_BATCH_NUM = 1024
```
[the remark this answers is not in the document]
Comment (yuxianq) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py]: > Can this cause memory overhead if many micro batches are in flight?

Yes, it will only add slight host memory overhead to store in-flight batches. No GPU memory overhead.

> I guess the total number of requests is still limited by [this](https://github.com/yuxianq/TensorRT-LLM/blob/48a7116b9a284779cf4a1a8dd771cf4c98456c22/tensorrt_llm/_torch/pyexecutor/_util.py#L749), right?

Yes, it won't affect the scheduler's capacity.

File: tensorrt_llm/_torch/pyexecutor/py_executor.py (hunk 5 above)
Comment (yuxianq) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py]: Sure, how about `_ring_broadcast_sample_state`?

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 8]
```diff
@@ -1138,115 +1191,218 @@ def _executor_loop_pp(self):
...
+            self._disagg_pp_termination_handler.terminate_pending_requests()
+
+        if self.enable_iter_perf_stats and previous_batch is not None:
+            sample_state = previous_batch.sample_state
+            sample_state.scheduled_requests.context_requests = previous_batch.scheduled_ctx_reqs
+            self._process_iter_stats(
+                finished_requests,
+                self.active_requests,
+                previous_batch,
+                previous_batch.microbatch_id % self.dist.pp_size,
```
[the remark this answers is not in the document]
Comment (yuxianq) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py]: It is a fix for test. I think the user still expect to see `microbatch_id` in range of `[0, pp_size)`. Allocating more mocrobatch slots for async sample state broadcast is just an implementation detail. This modulo keeps the same behavior before/after this PR for stats.

File: tests/unittest/llmapi/test_llm_multi_gpu_pytorch.py [hunk 9]
```diff
@@ -140,7 +140,8 @@ def test_llm_return_logprobs_streaming_tp2(prompt_logprobs, logprobs,
     ],
 )
 def test_llm_get_stats_pp2(return_context_logits, enable_chunked_prefill,
-                           enable_iter_req_stats):
+                           enable_iter_req_stats, monkeypatch):
+    monkeypatch.setenv("TLLM_PP_ASYNC_BROADCAST_SAMPLE_STATE", "0")
```
[the remark this answers is not in the document]
Comment (yuxianq) [reply tests/unittest/llmapi/test_llm_multi_gpu_pytorch.py]: Moved it into `llm_get_stats_test_harness` and added comments.

File: tensorrt_llm/_torch/pyexecutor/py_executor.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (yuxianq) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py]: Added comments.

File: tensorrt_llm/_torch/pyexecutor/py_executor.py (hunk 5 above)
Comment (yuxianq) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py]: Updated.

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 10]
```diff
@@ -1138,115 +1191,218 @@ def _executor_loop_pp(self):
...
+                else:
+                    for executed_batch in executed_batches:
+                        self._handle_executed_batch(executed_batch)
+                self.unhandled_batch_counter -= executed_batch_num
 
-                        self._handle_logits_communication(
-                            previous_batch, prev_microbatch_id)
+                # march forward in microbatch slots
+                if can_queue:
+                    self.previous_batch = batch_state
```
[the remark this answers is not in the document]
Comment (yuxianq) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py]: I have moved it before stage 2.
We cannot set `self.previous_batch = None` if `can_queue` is false since we need to record the queued batch as previous batch. See explanation above.

File: tensorrt_llm/llmapi/mpi_session.py [hunk 11]
```diff
@@ -170,8 +170,14 @@ def abort(self):
...
 
+        env = {
+            key: value
+            for key, value in os.environ.items()
+            if key.startswith("TRTLLM") or key.startswith("TLLM")
+        }
         self.mpi_pool = MPIPoolExecutor(max_workers=self.n_workers,
-                                        path=sys.path)
+                                        path=sys.path,
+                                        env=env)
```
Comment (syuoni) [inline tensorrt_llm/llmapi/mpi_session.py]: I was also previously thinking of implementing this. Thanks! cc @Superjomn

**Review by syuoni — APPROVED** [review]
Approved for the LLM API part, thanks!

**Review by Funatiq — APPROVED** [review]
